### PR TITLE
#258 add duration types to general component

### PR DESCRIPTION
**Pull request by SlakvichAleksandr**

### CODE REVIEW

**Comment by SlakvichAleksandr** [comment]
https://prnt.sc/Qm9dTkStynLL
https://prnt.sc/eQMKZn9llckb
https://prnt.sc/ibdC-NALM7Qh

**Comment by SlakvichAleksandr** [comment]
https://user-images.githubusercontent.com/78310758/189070342-a05b751f-1d82-48f6-a038-8e903652a2c2.mp4

**Review by Ymkavi — APPROVED** [review]

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 1]
```diff
@@ -149,4 +159,28 @@ export class GeneralComponent implements OnInit {
     colorInputChanged(color: Color) {
         this.settings.color = color;
     }
+
+    onDurationChange() {
+        this.customTimeShown = this.duration.time === 'Custom';
+        this.settings.size = this.duration.minutes!;
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: please get rid of mutation

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 2]
```diff
@@ -149,4 +159,28 @@ export class GeneralComponent implements OnInit {
...
+    }
+
+    customDurationChanged() {
+        const unitOfTime: UnitOfTime = UnitOfTime[this.selectedUnit as keyof typeof UnitOfTime];
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: why cannot use just this.selectedUnit

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 3]
```diff
@@ -149,4 +159,28 @@ export class GeneralComponent implements OnInit {
...
+
+        this.settings.size =
+            unitOfTime === UnitOfTime.Hour
+                ? parseInt(this.inputCustomTime, 10) * 60
+                : parseInt(this.inputCustomTime, 10);
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: please a separate method to update settings size

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 4]
```diff
@@ -149,4 +159,28 @@ export class GeneralComponent implements OnInit {
...
+    }
+
+    private defineCurrentDuration(slotDuration: number) {
+        this.duration = this.durations.find((x) => x.minutes === slotDuration) ?? this.durations[4];
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: get rid of magic number

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 5]
```diff
@@ -149,4 +159,28 @@ export class GeneralComponent implements OnInit {
...
+
+        if (this.duration.time === 'Custom') {
+            this.customTimeShown = true;
+            this.selectedUnit = this.unitOfTime.find((x) => x === 'Min')!;
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: use enam

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 6]
```diff
@@ -53,6 +52,7 @@ export class GeneralComponent implements OnInit {
             this.meetingPaddings.push(this.advancedSettings.paddingMeeting);
         }
         this.addAdvanced = Boolean(this.slot?.advancedSlotSettingsId);
+        this.defineCurrentDuration(this.settings.size);
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts:44]: Please ger rid of push

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 7]
```diff
@@ -149,4 +159,33 @@ export class GeneralComponent implements OnInit {
...
+
+        if (this.duration.time === 'Custom') {
+            this.customTimeShown = true;
+            this.selectedUnit = this.unitOfTime.find((x) => x === String(UnitOfTime.Min))!;
+            this.inputCustomTime = String(slotDuration);
+        }
+    }
+
+    private updateSettingSize(slotSize: number) {
+        this.settings.size = slotSize;
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: Get rid of mutation

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 8]
```diff
@@ -149,4 +159,33 @@ export class GeneralComponent implements OnInit {
...
+        this.updateSettingSize(slotSize);
+    }
+
+    private defineCurrentDuration(slotDuration: number) {
+        this.duration =
+            this.durations.find((x) => x.minutes === slotDuration) ?? this.durations.find((x) => x.time === 'Custom')!;
+
+        if (this.duration.time === 'Custom') {
+            this.customTimeShown = true;
+            this.selectedUnit = this.unitOfTime.find((x) => x === String(UnitOfTime.Min))!;
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: Do not use '!'

Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: Use another way. Without String

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: frontend/src/app/modules/availability/new-slot/general/general.component.sass [hunk 9]
```diff
@@ -67,6 +67,23 @@ input[type=text], .mat-select, .label-container, .error
 .input-container
     width: 100%
 
+    .custom-time
+        display: flex
+        align-items: flex-start
+        justify-content: space-between
+        max-width: calc(280px - 20px)
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.sass]: please use % or em

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 10]
```diff
@@ -149,4 +147,43 @@ export class GeneralComponent implements OnInit {
...
     }
+
+    onDurationChange() {
+        this.customTimeShown = this.duration.time === 'Custom';
+        this.settings = this.updateSettingSize(this.duration.minutes!);
+    }
+
+    customDurationChanged() {
+        const slotSize: number =
+            UnitOfTime[this.selectedUnit as keyof typeof UnitOfTime] === UnitOfTime.Hour
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]: why you cannot just use this.selectedUnit as UnitOfTime?

File: frontend/src/app/modules/availability/new-slot/general/general.component.ts [hunk 11]
```diff
@@ -149,4 +147,43 @@ export class GeneralComponent implements OnInit {
...
+    private updateSettingSize(slotSize: number): ISaveGeneralSettings {
+        const settings: ISaveGeneralSettings = {
+            size: slotSize,
+            isVisible: this.settings.isVisible,
+            color: this.settings.color,
+            name: this.settings.name,
+            locationType: this.settings.locationType,
+            isEnabled: this.settings.isEnabled,
+            type: this.settings.type,
+        };
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/general/general.component.ts]:
```suggestion
        const settings: ISaveGeneralSettings = {
            ...this.settings,
            size: slotSize
        };
```

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: frontend/src/app/modules/availability/new-slot/general/general.component.sass (hunk 9 above)
Comment (tatianahutii) [reply frontend/src/app/modules/availability/new-slot/general/general.component.sass]: and I believe we don't need calc here =)

**Review by tatianahutii — APPROVED** [review]

**Review by apofis96 — APPROVED** [review]

**Pull request merged**